src/helpers/mock.rs

```diff
@@ -0,0 +1,286 @@
+                    match msg {
+                        ControlMessage::ConnectionRequest(peer, step, peer_connection) => {
+                            controller.connect(peer, step, peer_connection);
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    fn connect(&self, peer: Identity, step: S, mut rx: Receiver<MessageEnvelope>) {
+        assert_ne!(self.identity, peer);
+
+        // start listening for incoming messages and move them from channel to the buffer
+        tokio::spawn({
+            let buf = Arc::clone(&self.buf);
+            async move {
+                while let Some(msg) = rx.recv().await {
+                    let mut buf = buf.lock().unwrap();
+                    buf.entry((peer, step)).or_insert_with(Vec::new).push(msg);
+                }
+            }
+        });
+    }
+
+    async fn get_connection(&self, peer: Identity, step: S) -> Sender<MessageEnvelope> {
+        assert_ne!(self.identity, peer);
+
+        loop {
+            // Depending on connection status, request a new connection, spin and wait for
+            // connection acknowledgment from peer or return the sender end of connection
+            // if it is ready
+            let control_message = {
+                let mut connections = self.connections.lock().unwrap();
+                let conn_state = connections
+                    .entry((peer, step))
+                    .or_insert_with(|| ConnectionState::Listen);
+
+                match conn_state {
+                    ConnectionState::Listen => {
+                        let (tx, rx) = channel(1);
+                        *conn_state = ConnectionState::Established(tx);
+
+                        Some(ControlMessage::ConnectionRequest(self.identity, step, rx))
+                    }
+                    ConnectionState::Established(sender) => {
+                        return sender.clone();
+                    }
+                }
+            };
+
+            if let Some(msg) = control_message {
+                self.peers
+                    .get(&peer)
+                    .unwrap_or_else(|| panic!("No peer with id {peer:?}"))
+                    .send(msg)
+                    .await
+                    .unwrap();
+            }
+            tokio::task::yield_now().await;
+        }
+    }
+
+    async fn receive(&self, peer: Identity, step: S, record_id: RecordId) -> MessageEnvelope {
+        // spin and wait until message with the same record id appears in the buffer
+        // when it happens, pop it out, try to reinterpret its bytes as `T` and return
+        loop {
+            {
+                let mut buf = self.buf.lock().unwrap();
+                if let Some(msgs) = buf.get_mut(&(peer, step)) {
+                    let l = msgs.len();
+                    for i in 0..l {
+                        if msgs[i].record_id == record_id {
+                            msgs.swap(i, l - 1);
```

Comment:
this works, but seems like it'd be easier to understand as a `HashMap<RecordId, MessageEnvelope>`

Reply:
yea, initially I thought that I should just have a linear buffer of objects but ended up implementing a lookup by `record_id` which is what hashmap is for. I'll probably keep it around as there are other ideas circulating about the internals so this method may be changed entirely

Reply:
👍